shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/HardFork.hs

```diff
@@ -0,0 +1,171 @@
+translateUTxOByronToShelley (Byron.UTxO utxoByron) =
+    UTxO $ Map.fromList
+      [ (txInShelley, txOutShelley)
+      | (txInByron, txOutByron) <- Map.toList utxoByron
+      , let txInShelley  = translateCompactTxInByronToShelley  txInByron
+            txOutShelley = translateCompactTxOutByronToShelley txOutByron
+      ]
+
+translateToShelleyLedgerState
+  :: forall sc. Crypto sc
+  => ShelleyGenesis sc
+  -> Globals
+  -> EpochNo
+  -> Byron.ChainValidationState
+  -> ShelleyState sc
+translateToShelleyLedgerState genesisShelley globals epochNo cvs =
+    NewEpochState {
+        nesEL     = epochNo
+      , nesBprev  = BlocksMade Map.empty
+      , nesBcur   = BlocksMade Map.empty
+      , nesEs     = epochState
+      , nesRu     = SNothing
+      , nesPd     = PoolDistr Map.empty
+      , nesOsched = oSchedule
+      }
+  where
+    pparams :: PParams
+    pparams = sgProtocolParams genesisShelley
+
+    -- | NOTE: we ignore the Byron delegation map because the genesis and
+    -- delegation verification keys are hashed using a different hashing
+    -- scheme, Blake2b_224, whereas Shelley uses Blake2b_256. This means we
+    -- can't simply convert them, as Byron nowhere stores the original
+    -- verification keys.
+    --
+    -- Fortunately, no Byron genesis delegations have happened yet, and if
+    -- they did, we would be aware of them before the hard fork, as we
+    -- instigate the hard fork. We just have to make sure that the hard-coded
+    -- Shelley genesis contains the same genesis and delegation verification
+    -- keys, but hashed with the right algorithm.
+    genDelegs :: GenDelegs sc
+    genDelegs = GenDelegs $ sgGenDelegs genesisShelley
+
+    reserves :: Coin
+    reserves =
+        fromIntegral (sgMaxLovelaceSupply genesisShelley) - balance utxoShelley
+
+    epochState :: EpochState sc
+    epochState = EpochState {
+        esAccountState = AccountState (Coin 0) reserves
+      , esSnapshots    = emptySnapShots
+      , esLState       = ledgerState
+      , esPrevPp       = pparams
+      , esPp           = pparams
+      , esNonMyopic    = emptyNonMyopic
+      }
+
+    utxoByron :: Byron.UTxO
+    utxoByron = Byron.cvsUtxo cvs
+
+    utxoShelley :: UTxO sc
+    utxoShelley = translateUTxOByronToShelley utxoByron
+
+    ledgerState :: LedgerState sc
+    ledgerState = LedgerState {
+        _utxoState = UTxOState {
+            _utxo      = utxoShelley
+          , _deposited = Coin 0
+          , _fees      = Coin 0
+          , _ppups     = emptyPPUPState
+          }
+      , _delegationState = DPState {
+          _dstate = emptyDState { _genDelegs = genDelegs }
+        , _pstate = emptyPState
+        }
+      }
+    oSchedule :: OverlaySchedule sc
+    oSchedule =
+      flip runReader globals $
+        overlaySchedule
+          epochNo
+          (Map.keysSet (sgGenDelegs genesisShelley))
+          (sgProtocolParams genesisShelley)
```

Comment:
Also incorporate the comments on the issue, e.g., https://github.com/input-output-hk/ouroboros-network/issues/2253#issuecomment-654842803.